File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution
{
    public TreeNode AddOneRow(TreeNode root, int val, int depth)
    {
        if (root == null) {
            return null;
        }
        if (depth == 1) 
        {
            return new TreeNode(val, root);
        }
        if (depth == 2)
        {
            root.left = new TreeNode(val, root.left, null);
            root.right = new TreeNode(val, null, root.right);
            return root;
        }
        AddOneRow(root.left, val, depth - 1);
        AddOneRow(root.right, val, depth - 1);
        return root;
    }
}

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        # if root is None:
        #     return
        
        if depth == 1:
            return TreeNode(val, root, None)
        
        # if depth == 2:
        #     root.left = TreeNode(val, root.left, None)
        #     root.right = TreeNode(val, None, root.right)
        #     return root
        
        # self.addOneRow(root.left, val, depth - 1)
        # self.addOneRow(root.right, val, depth - 1)
        # return root
        
        level = [root]
        
        for _ in range(2, depth):
            level = [child for node in level for child in (node.left, node.right) if child is not None]
        
        for node in level:
            node.left = TreeNode(val, node.left, None)
            node.right = TreeNode(val, None, node.right)
        
        return root
